ext/search_kit/Civi/Api4/Action/SearchDisplay/Download.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+namespace Civi\Api4\Action\SearchDisplay;
+
+/**
+ * Download the results of a SearchDisplay as a spreadsheet.
+ *
+ * Note: unlike other APIs this action directly outputs a file.
+ *
+ * @package Civi\Api4\Action\SearchDisplay
+ */
+class Download extends AbstractRunAction {
+
+  /**
+   * Requested file format
+   * @var string
+   * @required
+   * @options csv
+   */
+  protected $format;
+
+  /**
+   * @param \Civi\Api4\Generic\Result $result
+   * @throws \API_Exception
+   */
+  protected function processResult(\Civi\Api4\Generic\Result $result) {
+    $entityName = $this->savedSearch['api_entity'];
+    $apiParams =& $this->savedSearch['api_params'];
+    $settings = $this->display['settings'];
+
+    // Displays are only exportable if they have actions enabled
+    if (empty($settings['actions'])) {
+      \CRM_Utils_System::permissionDenied();
+    }
+
+    // Force limit if the display has no pager
+    if (empty($settings['pager']) && !empty($settings['limit'])) {
+      $apiParams['limit'] = $settings['limit'];
+    }
+    $apiParams['orderBy'] = $this->getOrderByFromSort();
+    $this->augmentSelectClause($apiParams);
+
+    $this->applyFilters();
+
+    $apiResult = civicrm_api4($entityName, 'get', $apiParams);
+
+    $rows = $this->formatResult($apiResult);
+
+    $columns = [];
+    foreach ($this->display['settings']['columns'] as $col) {
+      $col += ['type' => NULL, 'label' => '', 'rewrite' => FALSE];
+      if ($col['type'] === 'field' && !empty($col['key'])) {
+        $columns[] = $col;
+      }
+    }
+
+    // This weird little API spits out a file and exits instead of returning a result
+    \CRM_Utils_System::setHttpHeader('Content-Type', 'text/csv');
+    \CRM_Utils_System::setHttpHeader('Content-Disposition', 'attachment; filename=' . \CRM_Utils_String::munge($this->display['label']) . '.' . $this->format);
+    \CRM_Utils_System::setHttpHeader('Pragma', 'no-cache');
+    \CRM_Utils_System::setHttpHeader('Expires', '0');
+
+    $out = fopen('php://output', 'w');
+    fputcsv($out, array_column($columns, 'label'));
```

Comment:
I haven't tested this but there was a series of tickets/PRs about utf-8. I expect it needs the BOM.

Here's one: https://github.com/civicrm/civicrm-core/pull/15968

Reply:
Tried an export here on the test site. Open in excel without BOM:

<img width="168" alt="Untitled" src="https://user-images.githubusercontent.com/2967821/131525918-e20f7b13-dfc0-4db1-bd50-efb4592f6296.png">

After I added the BOM to the file locally:

<img width="164" alt="Untitled2" src="https://user-images.githubusercontent.com/2967821/131525959-b8fb98a7-c9ef-430a-9ddc-511f90ac7571.png">

Reply:
Thanks @demeritcowboy I've added that to the start of the output.

Reply:
Hmm, doesn't seem to work. Just double-checking something.

Reply:
Ok because of the method being used here to output to the browser, it will strip out the BOM, so a workaround is to output it twice, i.e. `echo "\xEF\xBB\xBF\xEF\xBB\xBF";` This feels super-silly though.

Reply:
Are you sure? At lease in LibreOffice (I don't have MS Office) I verified that it was getting the encoding wrong before, and your suggested addition fixed it for me.

Reply:
I double-checked now with both chrome and firefox and the file itself is missing the BOM on download. Maybe OS+browser matters? Hmm.

Reply:
Does using fwrite instead of echo help at all?
```diff
diff --git a/ext/search_kit/Civi/Api4/Action/SearchDisplay/Download.php b/ext/search_kit/Civi/Api4/Action/SearchDisplay/Download.php
index 4e4d3f736f..68ced867ab 100644
--- a/ext/search_kit/Civi/Api4/Action/SearchDisplay/Download.php
+++ b/ext/search_kit/Civi/Api4/Action/SearchDisplay/Download.php
@@ -79,9 +79,9 @@ class Download extends AbstractRunAction {
     // Output UTF BOM so that MS Excel copes with diacritics. This is recommended as
     // the Windows variant but is tested with MS Excel for Mac (Office 365 v 16.31)
     // and it continues to work on Libre Office, Numbers, Notes etc.
-    echo "\xEF\xBB\xBF";
 
     $out = fopen('php://output', 'w');
+    fwrite($out, "\xEF\xBB\xBF");
     fputcsv($out, array_column($columns, 'label'));
 
     foreach ($rows as $data) {

```

Reply:
Good thought but no. I also tested a civireport export just in case maybe that no longer works either but it does. I also tried reworking this to use 'php://temp' and then just a pure echo at the end reading from the stream, but same output. I'm confused because that's very similar to the civireport.

Reply:
@colemanw the best way to output a csv is to use the League package that we already used in core - it deals with a bunch of gotchas like BOMs

Reply:
This is what we have been migrating our code to use https://csv.thephpleague.com/

Reply:
(I didn't think to mention it earlier cos I thought you were aware of it)

Reply:
I'm thinking the issue here is something else, since it doesn't explain the civireport which works, but yeah might be another thing to try.
There's a progress bar here but I'm not sure it's the problem either, just something different than the civireport.

Reply:
FWIW this bit I stole off the internet works, suggesting it is kinda related to the progress bar or at least the js blob concept.

```patch
--- a/ext/search_kit/ang/crmSearchTasks/crmSearchTaskDownload.ctrl.js
+++ b/ext/search_kit/ang/crmSearchTasks/crmSearchTaskDownload.ctrl.js
@@ -26,7 +26,13 @@
         params: JSON.stringify(apiParams)
       })
         .done(function(data, status, jqXHR) {
-          var blob = new Blob([data], {
+          var buffer = new ArrayBuffer(3);
+          var dataView = new DataView(buffer);
+          dataView.setUint8(0, 0xef);
+          dataView.setUint8(1, 0xbb);
+          dataView.setUint8(2, 0xbf);
+          var bom = new Uint8Array(buffer);
+          var blob = new Blob([bom, data], {
             type: jqXHR.getResponseHeader('Content-Type')
           }),
             a = document.createElement("a"),
```

Reply:
Thanks @demeritcowboy that pointed me in the right direction. The issue seems to be that jQuery `$.post` doesn't support ArrayBuffer but angular's `$http.post` does. It came with its own set of problems but I think I've worked around them.

I've rebased this PR. Would you be able to test again?

Reply:
Hmm doesn't seem to do it. When you download the file do you get the BOM? `od -x -c file.csv | head` is a good way to see if it has the three bytes at the start (note the hex part shows the byte pairs in backwards order, but it's useful to see the byte value).